README.md
[![Clojars Project](https://img.shields.io/clojars/v/jmh-clojure/task.svg)](https://clojars.org/jmh-clojure/task)
[![](https://github.com/jgpc42/jmh-clojure-task/workflows/Test%20runner/badge.svg)][ci]

### Dependency and version information
<details>
  <summary>Click to show</summary>

[Leiningen][lein]

``` clojure
[jmh-clojure/task "0.1.1"]
```

[tools.deps][deps]

```clojure
{jmh-clojure/task {:mvn/version "0.1.1"}}
```

[Maven](http://maven.apache.org)

``` xml
<dependency>
  <groupId>jmh-clojure</groupId>
  <artifactId>task</artifactId>
  <version>0.1.1</version>
</dependency>
```

JDK versions 8 to 16 and Clojure versions 1.7 to 1.10 are currently [tested against][ci].
</details>

### What is it?

Various convenience utilities for [jmh-clojure][jmh-clj] intended to be used by higher-level tools. Note that while the examples below use `tools.deps`, this library can also be used via a [`Leiningen plugin`][lein-jmh], or standalone, as well.

### Usage

For `clj`, the recommend way to use this library is by adding an alias to your `deps.edn`. The following will be sufficient for most users:

```clojure
{#_...
 :aliases
 {:jmh {:extra-paths ["classes"]
        :extra-deps {jmh-clojure/task {:mvn/version "0.1.1"}}
        :ns-default jmh.task
        :exec-fn jmh.task/run}}}
```

It's important to notice that the `"classes"` directory is added to the project paths. This is required for `jmh-clojure` since it generates `.class` files dynamically (written to `*compile-path*`, by default). We must add this output directory so it's available on the classpath when running.

By default, a `jmh.edn` file at the root of your project is used to configure the benchmark runner. Please see the [sample file][sample] for a complete guide. For example, the following runs all benchmarks, enables the standard JMH status log, and outputs pretty-printed benchmark result data:

```bash
$ mkdir -p classes
$ clj -X:jmh :format :pprint, :status true
```

> If you're new to the `-X` flag, it allows us to omit the Clojure map curly braces (`{}`) on the command line.

Notice again that we must manually create the `classes` directory before running. Ensuring the `*compile-path*` directory exists is not automated by `tools.deps` as it is with other tools like Leiningen.

### Available options

The optional map to `jmh.task/run` specifies the normal jmh-clojure run [options][run-doc]. The extra task options provided by this library include:

| Option        | Description                                    |
| ------------- | ---------------------------------------------- |
| `:exclude`    | keys to remove from each benchmark result map. |
| `:file`       | read the given file(s) instead of `jmh.edn`.   |
| `:format`     | print results in the given format(s).          |
| `:only`       | keys to select from each benchmark result map. |
| `:output`     | specify the location to write the results.     |
| `:pprint`     | equivalent to `:format :pprint`.               |
| `:progress`   | display progress data while running.           |
| `:sort`       | key(s) to order the results by.                |

Please run `clj -X:jmh help` for more information on the available options. Note that some formats (e.g., `:table`) exclude some result information due to space considerations. See the previously mentioned help for details.

The available JMH profilers may be listed with: `clj -X:jmh profilers`.

### Uberjar example

Along with integrating with the Clojure project tool of your choice, you can also create a standalone jar file. This is useful if you intend to run benchmarks on a different machine than your development one and want to package everything up so you don't have to re-create the project build environment. The `jmh.main` namespace is designed to be used as the jar entry point.

For Clojure tools, we'll use the [`uberdeps`][udeps] library to create the jar. In the root of a hypothetical project:

```bash
mkdir -p classes uberdeps
echo '{:deps {uberdeps/uberdeps {:mvn/version "1.0.4"}}}' > uberdeps/deps.edn
clj -Ajmh -M -e "(compile 'jmh.main)"
cd uberdeps
clj -M -m uberdeps.uberjar --aliases jmh --deps-file ../deps.edn --main-class jmh.main --target ../target/jmh.jar
mkdir classes
cp ../jmh.edn .
java -cp classes:../target/jmh.jar jmh.main '{:type :quick}'
```

Note again that even when running as an uberjar, we still need to create the `classes` directory beforehand to enable the dynamic compilation. Additionally, we copy `jmh.edn` to the working directory so it can be found by the runner. This file can also be packaged as a jar resource so this step is optional. Running with `-Dfile.encoding=UTF-8` is also advisable depending on your platform due to the unicode characters JMH can output.

The procedure for other tools should be similarly straightforward.

### More information

Please see the [`jmh-clojure`][jmh-clj] project for everything else about jmh-clojure.

### Running the tests

```bash
lein test
```

Or, `lein test-all` for all supported Clojure versions.

### License

Copyright © 2017-2021 Justin Conklin

Distributed under the Eclipse Public License, the same as Clojure.



[ci]:        https://github.com/jgpc42/jmh-clojure-task/blob/master/.github/workflows/test.yml
[deps]:      https://github.com/clojure/tools.deps.alpha
[jmh-clj]:   https://github.com/jgpc42/jmh-clojure
[lein]:      http://github.com/technomancy/leiningen
[lein-jmh]:  https://github.com/jgpc42/lein-jmh
[run-doc]:   https://jgpc42.github.io/jmh-clojure/doc/jmh.core.html#var-run
[sample]:    https://github.com/jgpc42/jmh-clojure/blob/master/resources/sample.jmh.edn
[udeps]:     https://github.com/tonsky/uberdeps


CHANGELOG.md
## 0.1.1 (2021-08-25)

* A few additions to make running via `clj` more idiomatic

> **Note:** this repo was extracted from the [`lein-jmh`][origin] project.



[origin]:  https://github.com/jgpc42/lein-jmh/tree/84b0350/task


src/jmh/task.clj
(ns jmh.task
  "Utilities for jmh-clojure."
  (:require [jmh.core :as jmh]
            [clojure.edn :as edn]
            [clojure.java.io :as io]
            [clojure.set :as set]
            [clojure.pprint :as pprint :refer [pprint]])
  (:import [java.nio.file Files FileSystems FileVisitOption Path Paths]
           [java.util.function Predicate]
           [org.openjdk.jmh.util ScoreFormatter]))

(defmulti report
  "Write the given benchmark results in the specified format."
  (fn [format result] format)
  :default :default)

(defn- format-score [v]
  (ScoreFormatter/format v))

(defn- format-score-error [v]
  (ScoreFormatter/formatError v))

(defn- keyword-seq
  "Coerce value to a sequence of keywords."
  [x]
  (seq (if (keyword? x) [x] x)))

(defn- some-update
  "Update the given key in the map if it is truthy."
  [m k f & args]
  (apply update m k #(and % (apply f % %&)), args))

(defn merge-recursively
  "Merge two like-values recursively with an appropriate combining fn.
  If both values are not maps, sets, or collections, return the second."
  [a b]
  (cond
    (and (map? a) (map? b))
    (merge-with merge-recursively a b)
    (and (set? a) (set? b))
    (set/union a b)
    (and (coll? a) (coll? b))
    (concat a b)
    :else
    b))

(defn- read-file
  "Read the given jmh environment file as edn data."
  [f]
  (try
    (-> f slurp edn/read-string)
    (catch Exception e
      (condp = (.getMessage e)
        "No dispatch macro for: ("
        (throw (RuntimeException.
                (str "edn format does not support the '#(...)' "
                     "anonymous function reader macro")
                e))
        "No dispatch macro for: \""
        (throw (RuntimeException.
                (str "edn format does not support the '#\"...\"' "
                     "regex reader macro")
                e))
        (throw e)))))

(defn glob
  "Return a File seq of paths matching the wildcard pattern."
  [opts ^String pattern]
  (when pattern
    (let [root (:jmh.task/root opts ".")
          path (Paths/get root (into-array String []))
          matcher (-> (FileSystems/getDefault)
                      (.getPathMatcher (str "glob:" pattern)))
          paths (-> (Files/walk path (into-array FileVisitOption []))
                    (.filter (reify Predicate
                               (test [_ p] (.matches matcher (.relativize path ^Path p)))))
                    .toArray)]
      (map #(.toFile ^Path %) paths))))

(defn merge-environment
  "Return the jmh environment map."
  [opts]
  (let [root (:jmh.task/root opts ".")
        files (mapcat
               (fn [x]
                 (cond
                   (string? x)
                   (let [f (io/file root x)]
                     (when (.exists f)
                       [f]))
                   (and (vector? x) (= (first x) :resource))
                   (when-let [resx (io/resource (second x))]
                     [resx])
                   (and (vector? x) (= (first x) :glob))
                   (glob opts (second x))
                   :else
                   nil))
               (:files opts))
        envs (map read-file files)]
    (reduce merge-recursively nil
            (concat envs [(:jmh.task/env opts {})]))))

(def ^{:dynamic true
       :doc "The default :files that provide the benchmark environment."}
  *files* ["jmh.edn" [:resource "jmh.edn"]])

(def ^{:doc "The default options for each format."}
  format-options
  {:table {:exclude [:score-confidence :statistics :threads]}})

(declare progress-reporter)

(defn normalize-options
  "Return the merged options map for the task options."
  [opts]
  (let [opts (if (:pprint opts)
               (assoc opts :format :pprint)
               opts)

        file (:file opts (:files opts *files*))
        files (if (and (coll? file) (not (keyword? (first file))))
                file
                [file])
        formats (keyword-seq (:format opts :default))

        defs (apply merge (for [f formats] (get format-options f)))
        opts (merge defs opts)

        p (:progress opts :out)
        f (if (or (#{:out :err} p)
                  (true? p))
            (progress-reporter (if (= p :err) *err* *out*))
            (eval p))]

    (-> (assoc opts :files files :formats formats :progress f)
        (update :exclude (comp set keyword-seq))
        (update :only keyword-seq)
        (update :sort keyword-seq))))

;;;

(defn format-table
  "Align the given column key values of the rows in a tabular format.
  Returns a string."
  [cols rows]
  (let [cols (filter #(some % rows) cols)

        widths (for [k cols]
                 (let [lens (map (comp count str k) rows)]
                   (apply max (cons (count (str k)) lens))))

        fmts (map #(format "~%da" %) widths)
        fmt (str "~:{~&"
                 (apply str (interpose "  " fmts))
                 "~}")

        seps (for [w widths]
                (apply str (repeat w \-)))
        header [cols seps]]

    (if (seq cols)
      (->> rows
           (map (apply juxt cols))
           (map #(replace {nil ""} %))
           (concat header)
           (pprint/cl-format nil fmt))
      "")))

(defn progress-reporter
  "Return a fn that will write events to the given Writer object."
  [out]
  (let [prev (volatile! nil)]
    (fn [{eta :eta, pct :percent}]
      (let [eta (long eta)
            pct (double pct)
            line (format "{:%% %.1f :eta \"%02d:%02d:%02d\"}"
                         (* 100.0 pct)
                         (long (/ eta (* 60.0 60)))
                         (long (/ (double (mod eta (* 60.0 60))) 60))
                         (mod eta 60))
            nline (count line)
            nprev (count @prev)
            prev? (pos? nprev)
            clear (if (and prev? (< nline nprev))
                    (apply str (repeat (- nprev nline) \space))
                    "")]
        (binding [*out* out]
          (locking out
            (when prev?
              (print \return))
            (print line)
            (print clear)
            (when (>= pct 1.0)
              (newline))
            (flush)
            (vreset! prev line)))))))

(defn result-comparator
  "Return a comparator fn that will sort results by the given keys."
  [xs]
  (let [xs (if (#{:asc :desc} (second xs))
             [xs]
             xs)]
    (fn [a b]
      (reduce
       (fn [n x]
         (let [[k order] (if (keyword? x)
                           [x :asc]
                           x)
               n (if (= order :desc)
                   (compare (k b) (k a))
                   (compare (k a) (k b)))]
           (if (zero? n)
             n
             (reduced n))))
       0 xs))))

;;;

(defn prepare-result
  "Sort and select result keys."
  [result opts]
  (cond->> result
    (:sort opts)
    (sort (result-comparator (:sort opts)))
    (:only opts)
    (map #(select-keys % (:only opts)))
    (not (:only opts))
    (map #(select-keys % (remove (:exclude opts) (keys %))))))

(defn run-benchmarks
  "Run the given benchmark environment and options."
  [env {dest :output :as opts}]
  (let [result (try
                 (jmh/run env opts)
                 (catch Exception e
                   (if (re-find #"^no benchmarks defined" (or (.getMessage e) ""))
                     (prn {:error (.getMessage e)
                           :files (:files opts)})
                     (.printStackTrace e))
                   (System/exit 1)))

        result (prepare-result result opts)

        output (with-out-str
                 (doseq [kind (:formats opts)]
                   (println)
                   (report kind result)))]

    (if (string? dest)
      (spit dest output)
      (binding [*out* (if (= :err dest) *err* *out*)]
        (print output)
        (flush)))))

(defn list-profilers
  "Print the available profilers as a table."
  []
  (let [rows (->> (jmh/profilers)
                  (filter :supported)
                  (sort-by :name))]
    (println (format-table [:name :desc] rows))))

(defn profilers [_]
  "Convenience entry point for `list-profilers`."
  (list-profilers))

(defn show-help
  "Print a general help message"
  []
  (println
   "Arguments: ([options-or-keyword] [])

  By default, a 'jmh.edn' file is used to specify the benchmark
  environment. It may be found in the current directory and/or available
  as a resource. This can be altered via the :file option, below.

  Normally, a jmh-clojure options map is provided as the task argument,
  in this case, the following options are recognized or overridden by
  this task runner:

  :exclude   the keys to remove of each result map. Overridden by :only.

  :file      specify another file to read instead of 'jmh.edn'. Can also
             be a tuple of [:glob \"pattern\"] or [:resource \"path\"].
             The ordering of globbed files may be undefined. See :files.

  :files     synonym for :file. A sequence of :file values. Each file's
             data is recursively merged from left to right.

  :format    keyword or sequence of keywords (for multiple outputs).
             See below. Unspecified or :default for normal output.

  :only      the keys to select of each result map. Overrides :exclude.

  :output    by default the :format results are written to stdout. If
             :err, the results are instead written to stderr. If a
             string is provided, write to the specified file.

  :progress  if true or :out, report progress to stdout, if :err report
             to stderr. If false, ignore. Defaults to :out.

  :pprint    equivalent to :format :pprint.

  :sort      key or key seq. Sort results by the given keys. Each key
             may also be a tuple of [key order], where order is either
             :desc or :asc (default).

  Available output formats:

  :pprint  pretty print results via `clojure.pprint/pprint`.

  :table   tabular format. Results with nested maps (e.g., :percentiles)
           are expanded. Due to width constraints, this format elides
           some information present in other formats. This format also
           excludes some keys automatically. Specify an empty collection
           to the :exclude option to show all columns.

  Instead of an options map, the following command keywords are available:

  :profilers  data about available profilers.
  :help       show this help message.

  Please see the jmh-clojure project for more information on
  configuration and options."))

(defn help [_]
  "Convenience entry point for `show-help`."
  (show-help))

(defn main
  "Run a jmh task. Accepts an optional command keyword or option map."
  ([] (main nil))
  ([task-arg] (main task-arg {}))
  ([task-arg extra-opts]
   (condp = task-arg
     :profilers (list-profilers)
     :help (show-help)
     (let [opts (-> (merge task-arg extra-opts)
                    normalize-options)
           env (merge-environment opts)]
       (run-benchmarks env opts)))))

(defn run [env]
  "Convenience entry point for the main jmh runner."
  (main env))

;;;

(defmethod report :default [_ result]
  (prn result))

(defmethod report :pprint [_ result]
  (pprint result))

;;;

(def ^{:private true
       :doc "The common table column for :fn and :method keys."}
  combine-key :benchmark)

(def ^{:private true
       :doc "The result keys that should be expanded."}
  expand-keys [:percentiles :secondary])

(defn pr-str-max
  "Convert `val` to string via `pr-str`. If the string is longer than
  `max-width`, truncate the middle and replace with `replacement`."
  ([val] (pr-str-max val 48 " ... "))
  ([val ^long max-width replacement]
   {:pre [(< (count replacement) max-width)]}
   (let [s (pr-str val)
         len (count s)]
     (if (> len max-width)
       (let [rlen (count replacement)
             off (quot rlen 2)
             mid (quot max-width 2)
             beg (dec (+ (- mid off) (rem max-width 2)))
             end (dec (+ (- len mid) off (rem rlen 2)))]
         (str (subs s 0 beg)
              replacement
              (subs s end len)))
       s))))

(defn align-column
  "Update each row by aligning the given tuple key's values in its
  column: the first to the left and the second to the right."
  [col-key sep-width f1 f2 rows]
  (let [rows (for [m rows]
               (if-let [[a b] (col-key m)]
                 (assoc m col-key [(f1 a) (f2 b)])
                 m))
        vals (map (comp count first col-key) rows)
        max (long (apply max 0 vals))]
    (for [m rows]
      (if-let [[a b] (col-key m)]
        (let [nspace (+ (long sep-width) (- max (count a)))
              sep (apply str (repeat nspace \space))]
          (assoc m col-key (str a sep b)))
        m))))

(defn format-statistics
  "Return a string representing some of the statistical data."
  [m]
  (format "{:min %s :max %s :stdev %s}"
          (format-score (:min m))
          (format-score (:max m))
          (format-score-error (:stdev m))))

(defn normalize-row
  "Prepare the result map for table display."
  [row]
  (let [b (:fn row (:method row))
        b (if (seq? b) (pr-str-max b) b)]
    (-> (dissoc row :fn :method)
        (assoc combine-key b)
        (some-update :params pr-str-max)
        (some-update :score-error format-score-error)
        (some-update :statistics format-statistics))))

(defn expand-rows
  "Flatten and append nested result map rows."
  [row]
  (let [expand
        (fn [row-key]
          (for [[k v] (row-key row)]
            (let [b (str "  " k
                         (when (number? k) "%"))]
              (-> (dissoc v :statistics)
                  (assoc combine-key b)
                  (some-update :score-error format-score-error)))))]
    (-> (apply dissoc row expand-keys)
        (cons (mapcat expand expand-keys)))))

(defmethod report :table [_ result]
  (let [cols [combine-key :name :mode :samples
              :score :score-error :score-confidence
              :threads :params :statistics]

        rows (->> (map normalize-row result)
                  (mapcat expand-rows)
                  (align-column :score 2 format-score identity)
                  (align-column :score-confidence 2 format-score format-score))]

    (println (format-table cols rows))))


project.clj
(def dependencies
  (->> "deps.edn" slurp read-string
       :deps (mapv #(vector (% 0) (:mvn/version (% 1))))
       (into '[[org.clojure/clojure "1.8.0"]])))

(defproject jmh-clojure/task "0.1.1"
  :description "Various jmh-clojure file and output utilities."
  :url "https://github.com/jgpc42/jmh-clojure-task"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}

  :dependencies ~dependencies

  :min-lein-version "2.0.0"

  :aliases {"test-all" ["do" "test,"
                        "with-profile" "+1.10" "test,"
                        "with-profile" "+1.9" "test,"
                        "with-profile" "+1.7" "test"]}

  :profiles
  {:1.7 {:dependencies [[org.clojure/clojure "1.7.0"]]}
   :1.9 {:dependencies [[org.clojure/clojure "1.9.0"]]}
   :1.10 {:dependencies [[org.clojure/clojure "1.10.0"]]}})
